File: shared/query/format/format_test.go
package format_test

import (
	"math/rand"
	"strconv"
	"strings"
	"testing"

	"github.com/vistasecurity/vistaplatform/shared/query/ast"
	"github.com/vistasecurity/vistaplatform/shared/query/format"
	"github.com/vistasecurity/vistaplatform/shared/query/parser"
)

func TestFormat(t *testing.T) {
	cases := map[string]string{
		// §10: implicit and written out, '-' written as not, ':=' as '='.
		"a:1 b:2":                        "a:1 and b:2",
		"-a:1":                           "not a:1",
		`a:="x"`:                         "a=x",
		"A:1 AND B:2":                    "a:1 and b:2",
		"a  :  1":                        "a:1",
		"a != 1":                         "a != 1",
		"a>=1":                           "a >= 1",
		"a in (1,2)":                     "a in (1, 2)",
		"a not in (1,2)":                 "a not in (1, 2)",
		"a:[1 to 2]":                     "a:[1 to 2]",
		`a~"x"`:                          `a ~ "x"`,
		"exists(a)":                      "exists(a)",
		"(a:1)":                          "a:1",
		"((a:1))":                        "a:1",
		"(a:1 or b:2)":                   "a:1 or b:2",
		"a:1 or (b:2)":                   "a:1 or b:2",
		"(a:1 or b:2) c:3":               "(a:1 or b:2) and c:3",
		"a:1 or b:2 and c:3":             "a:1 or b:2 and c:3",
		"not (a:1 or b:2)":               "not (a:1 or b:2)",
		"not (a:1 and b:2)":              "not (a:1 and b:2)",
		"a:(1 or 2)":                     "a:1 or a:2",
		"a:(1 and 2)":                    "a:1 and a:2",
		"a:(not 1)":                      "not a:1",
		"endpoint:(a:1)":                 "endpoint:(a:1)",
		"depends_on(1):(a:1)":            "depends_on:(a:1)",
		"depends_on(2):(a:1)":            "depends_on(2):(a:1)",
		"rel(connects_to, out, 1):(a:1)": "rel(connects_to, out):(a:1)",
		"last_seen<now-24h":              "last_seen < now-1d",
		"a:'x y'":                        `a:"x y"`,
	}
	for src, want := range cases {
		res, err := parser.Parse(src)
		if err != nil {
			t.Errorf("Parse(%q): %v", src, err)
			continue
		}
		if got := format.Format(res.Root); got != want {
			t.Errorf("Format(%q) = %q, want %q", src, got, want)
		}
	}
}

func TestFormatEmpty(t *testing.T) {
	if got := format.Format(nil); got != "" {
		t.Errorf("Format(nil) = %q, want the empty query", got)
	}
}

// TestFormatIdempotenceProperty is §10's property, exercised over a generated
// grammar rather than a fixed list: for any query the generator can produce,
// Format(Format(q)) == Format(q), and the canonical text parses to the same
// tree. A hand fuzzer rather than testing/quick because the interesting inputs
// are grammatical, and random bytes almost never are — FuzzParse covers those.
func TestFormatIdempotenceProperty(t *testing.T) {
	rng := rand.New(rand.NewSource(20260911))
	for i := 0; i < 3000; i++ {
		src := randomQuery(rng, 0)
		res, err := parser.Parse(src)
		if err != nil {
			t.Fatalf("generated query does not parse: %q: %v", src, err)
		}
		once := format.Format(res.Root)
		again, err := parser.Parse(once)
		if err != nil {
			t.Fatalf("canonical form %q of %q does not reparse: %v", once, src, err)
		}
		if twice := format.Format(again.Root); twice != once {
			t.Fatalf("not idempotent for %q:\n  once  %q\n  twice %q", src, once, twice)
		}
		if !ast.EqualJSON(res.Root, again.Root) {
			a, _ := ast.MarshalJSON(res.Root)
			b, _ := ast.MarshalJSON(again.Root)
			t.Fatalf("canonical form of %q parses differently:\n  %s\n  %s", src, a, b)
		}
	}
}

// randomQuery builds a grammatical query. depth bounds the recursion.
func randomQuery(rng *rand.Rand, depth int) string {
	if depth > 3 {
		return randomTerm(rng, depth)
	}
	switch rng.Intn(10) {
	case 0:
		return randomQuery(rng, depth+1) + " or " + randomQuery(rng, depth+1)
	case 1:
		return randomQuery(rng, depth+1) + " and " + randomQuery(rng, depth+1)
	case 2:
		return randomQuery(rng, depth+1) + " " + randomQuery(rng, depth+1)
	case 3:
		return "not " + randomQuery(rng, depth+1)
	case 4:
		return "-" + randomTerm(rng, depth+1)
	case 5:
		return "(" + randomQuery(rng, depth+1) + ")"
	default:
		return randomTerm(rng, depth)
	}
}

var (
	fields = []string{"hostname", "environment", "class", "risk_score", "attr.cpu_count",
		"tag.env", "id.mac", `tag."cost center"`, "last_seen"}
	values = []string{"production", "web-01", "70", "198.51.100.0/24", "aa:bb:*",
		`"two words"`, `"and"`, "now-24h", "now+90d", "3.0", "*", `"five*"`,
		// QUOTED strings shaped like a relative date. The formatter is
		// type-free, so it used to rewrite these too — see
		// TestQuotedDateShapedStringIsNotRewritten.
		`"now-24h"`, `"now"`, `"NOW+90D"`}
	rels  = []string{"depends_on", "used_by", "hosted_on", "any_rel"}
	colls = []string{"endpoint", "cert", "software", "finding"}
	ops   = []string{":", "=", "!=", " < ", " <= ", " > ", " >= "}
)

func randomTerm(rng *rand.Rand, depth int) string {
	switch rng.Intn(12) {
	case 0:
		return "exists(" + pick(rng, fields) + ")"
	case 1:
		return pick(rng, fields) + " in (" + pick(rng, values) + ", " + pick(rng, values) + ")"
	case 2:
		return pick(rng, fields) + " not in (" + pick(rng, values) + ")"
	case 3:
		return pick(rng, fields) + ":[" + strconv.Itoa(rng.Intn(50)) + " to " + strconv.Itoa(50+rng.Intn(50)) + "]"
	case 4:
		return pick(rng, fields) + ` ~ "^web[0-9]+"`
	case 5:
		return `"free text"`
	case 6:
		if depth > 2 {
			break
		}
		return pick(rng, colls) + ":(" + randomQuery(rng, depth+2) + ")"
	case 7:
		if depth > 2 {
			break
		}
		rel := pick(rng, rels)
		if rng.Intn(2) == 0 {
			rel += "(" + strconv.Itoa(1+rng.Intn(3)) + ")"
		}
		return rel + ":(" + randomQuery(rng, depth+2) + ")"
	case 8:
		if depth > 2 {
			break
		}
		return "rel(connects_to, " + pick(rng, []string{"out", "in", "any"}) + "):(" +
			randomQuery(rng, depth+2) + ")"
	case 9:
		return pick(rng, fields) + ":(" + pick(rng, values) + " or " + pick(rng, values) + ")"
	}
	return pick(rng, fields) + strings.TrimSpace(pick(rng, ops)) + pick(rng, values)
}

func pick(rng *rand.Rand, list []string) string { return list[rng.Intn(len(list))] }

// TestQuotedDateShapedStringIsNotRewritten is S7.
//
// §10 canonicalises a relative date into the largest exact unit, and the
// formatter is type-free — it decides by the shape of the node, never by a
// field's type. So it rewrote any value that LOOKED like one, quoted or not:
// `hostname="now-24h"` came out as `hostname=now-1d`, a different predicate
// that still parses. That is the worst shape a formatter bug can take, because
// nothing downstream can tell it happened.
//
// Quoting is exactly how a user says "I mean the text", so a quoted literal is
// never rewritten. The bare form carries no such ambiguity and still is.
func TestQuotedDateShapedStringIsNotRewritten(t *testing.T) {
	// The quotes are KEPT, for the same reason a quoted "*" keeps them:
	// unquoting would hand the bare form back to the date rule on the next
	// pass, so `Format(Format(q)) == Format(q)` would not hold.
	unchanged := map[string]string{
		`hostname="now-24h"`:      `hostname="now-24h"`,
		`hostname:"now-24h"`:      `hostname:"now-24h"`,
		`display_name:"now+90D"`:  `display_name:"now+90D"`,
		`hostname:'now-24h'`:      `hostname:"now-24h"`,
		`hostname in ("now-24h")`: `hostname in ("now-24h")`,
		// Free text is a substring search and can never be an instant, so it
		// is never date-canonicalised — quoted or not.
		`"now-24h"`: `now-24h`,
		`now-24h`:   `now-24h`,
		`"now+0Y"`:  `now+0Y`,
	}
	for src, want := range unchanged {
		res, err := parser.Parse(src)
		if err != nil {
			t.Errorf("Parse(%q): %v", src, err)
			continue
		}
		if got := format.Format(res.Root); got != want {
			t.Errorf("Format(%q) = %q, want %q — the stored predicate must not change", src, got, want)
		}
	}

	// The other polarity: a BARE date value is still canonicalised, or §10's
	// `24h` -> `1d` rule would have been dropped rather than narrowed.
	canonicalised := map[string]string{
		"last_seen<now-24h":              "last_seen < now-1d",
		"last_seen<now-14d":              "last_seen < now-2w",
		"last_seen < now-6months":        "last_seen < now-6mo",
		"last_seen:[now-24h to now+48h]": "last_seen:[now-1d to now+2d]",
	}
	for src, want := range canonicalised {
		res, err := parser.Parse(src)
		if err != nil {
			t.Errorf("Parse(%q): %v", src, err)
			continue
		}
		if got := format.Format(res.Root); got != want {
			t.Errorf("Format(%q) = %q, want %q", src, got, want)
		}
	}
}

File: packages/primitives/src/query/format.test.ts
// Canonical form (§10) as a property, not only as a table.
//
// The conformance file pins the canonical text of 222 specific queries. What it
// cannot show is that the two properties hold for queries nobody wrote down:
//
//   format(format(q)) === format(q)          — it is a normal form
//   parse(format(q))  ≡  parse(q)            — it is a normalisation, not a rewrite
//
// The second is the one that matters. A formatter that quietly changes what a
// query means while still producing text that parses is the worst shape this
// code can take — it would rewrite a stored scope into a different predicate
// and nothing would report an error.

import { describe, expect, it } from 'vitest';

import { format } from './format';
import { toJSON } from './json';
import { parse } from './parser';

function canonical(src: string): string {
  const out = parse(src);
  expect(out.ok, `parse ${JSON.stringify(src)}`).toBe(true);
  if (!out.ok) return '';
  return format(out.value.root);
}

/** A small grammar of query fragments, combined below into whole queries. */
const TERMS = [
  'environment:production',
  'hostname="Web01"',
  'hostname:=Web01',
  'risk_score >= 70',
  'risk >= high',
  'risk:not_assessed',
  'last_seen < now-24h',
  'last_seen < now-14d',
  'not_after < now+30d',
  'environment in (production, staging)',
  'environment not in (production, staging)',
  'risk_score:[40 to 69]',
  'hostname ~ "^web[0-9]{2}\\\\."',
  'exists(owner_email)',
  'owner_email:*',
  'payroll',
  '"has space"',
  '"now-24h"',
  'display_name:"five*"',
  'tag.env:prod',
  'tag:env',
  'id.mac:aa:bb:*',
  'id=550e8400-e29b-41d4-a716-446655440000',
  'class:hardware.computer.server',
  'class=server',
  'environment:(production or prod)',
  'environment:(not production and not staging)',
  'endpoint:(port:443 and protocol:tls)',
  'exists(cert)',
  'depends_on:(class=database_instance)',
  'depends_on(3):(display_name:"Payroll")',
  'depends_on(1):(class:server)',
  'used_by:(class:application)',
  'any_rel:(class:hardware)',
  'rel(connects_to, in, 2):(display_name:"core-sw-1")',
  'rel(connects_to, any):(class:server)',
  '-class:hardware',
  'not exists(owner_email)',
];

/** Combinators that must not change what the operands mean. */
function combinations(): string[] {
  const out: string[] = [...TERMS];
  for (let i = 0; i < TERMS.length; i++) {
    const a = TERMS[i];
    const b = TERMS[(i + 7) % TERMS.length];
    const c = TERMS[(i + 13) % TERMS.length];
    out.push(
      `${a} and ${b}`,
      `${a} or ${b}`,
      `${a} ${b}`, // juxtaposition is an implicit and (§2)
      `not (${a} or ${b})`,
      `not (${a} and ${b})`,
      `(${a} or ${b}) and ${c}`,
      `${a} or (${b} or ${c})`,
      `${a} and (${b} and ${c})`,
      `((${a}))`,
      `${a} and not ${b} or ${c}`,
    );
  }
  return out;
}

const QUERIES = combinations();

describe('canonical form is a normal form (§10)', () => {
  it('is idempotent for every generated query', () => {
    const notIdempotent: string[] = [];
    for (const q of QUERIES) {
      const once = canonical(q);
      const twice = canonical(once);
      if (once !== twice) notIdempotent.push(`${q}\n  once:  ${once}\n  twice: ${twice}`);
    }
    expect(notIdempotent).toEqual([]);
  });

  it('reparses to an identical tree for every generated query', () => {
    const changed: string[] = [];
    for (const q of QUERIES) {
      const before = parse(q);
      const after = parse(canonical(q));
      if (!before.ok || !after.ok) {
        changed.push(`${q}: canonical form does not parse`);
        continue;
      }
      const a = JSON.stringify(toJSON(before.value.root));
      const b = JSON.stringify(toJSON(after.value.root));
      if (a !== b) changed.push(`${q}\n  before: ${a}\n  after:  ${b}`);
    }
    expect(changed).toEqual([]);
  });

  it('covers a meaningful number of shapes', () => {
    // A property test over an empty corpus is the cheapest way to have a guard
    // that cannot fail.
    expect(QUERIES.length).toBeGreaterThan(300);
  });
});

describe('what canonical form normalises', () => {
  it('writes the implicit and, the dash negation and the := alias', () => {
    expect(canonical('environment:production class:server')).toBe(
      'environment:production and class:server',
    );
    expect(canonical('-class:hardware')).toBe('not class:hardware');
    expect(canonical('hostname:="Web01"')).toBe('hostname=Web01');
  });

  it('lowercases field names and keywords, and leaves values alone', () => {
    expect(canonical('Environment:Production AND Class:server')).toBe(
      'environment:Production and class:server',
    );
  });

  it('drops a redundant parenthesis and keeps a load-bearing one', () => {
    expect(canonical('(class:server)')).toBe('class:server');
    expect(canonical('(a:1 or b:2) and c:3')).toBe('(a:1 or b:2) and c:3');
    // An Or inside an Or would flatten into a different tree, so its
    // parentheses stay even though precedence does not require them.
    expect(canonical('a:1 or (b:2 or c:3)')).toBe('a:1 or (b:2 or c:3)');
  });

  it('writes a duration in the largest exact unit', () => {
    expect(canonical('last_seen < now-24h')).toBe('last_seen < now-1d');
    expect(canonical('last_seen < now-14d')).toBe('last_seen < now-2w');
    expect(canonical('last_seen < now-90d')).toBe('last_seen < now-90d');
    expect(canonical('last_seen < now-6months')).toBe('last_seen < now-6mo');
  });

  it('drops the default hop count and keeps any other', () => {
    expect(canonical('depends_on(1):(class:server)')).toBe('depends_on:(class:server)');
    expect(canonical('depends_on(3):(class:server)')).toBe('depends_on(3):(class:server)');
  });
});

describe('what canonical form deliberately does NOT do', () => {
  // §10: "Formatting is syntactic normalisation, not algebraic rewriting."
  it('preserves clause order', () => {
    expect(canonical('risk_score <= 10 and risk_score >= 1')).toBe(
      'risk_score <= 10 and risk_score >= 1',
    );
  });

  it('does not fold two comparisons into a range', () => {
    expect(canonical('risk_score >= 1 and risk_score <= 10')).toBe(
      'risk_score >= 1 and risk_score <= 10',
    );
  });

  it('does not rewrite a quoted value that looks like a date', () => {
    // The formatter is type-free: it cannot tell a date from a string that
    // looks like one, and `hostname="now-24h"` is a hostname. Rewriting it to
    // `hostname=now-1d` would change the stored predicate into a different one
    // that still parses (§13 S7).
    expect(canonical('hostname="now-24h"')).toBe('hostname="now-24h"');
    // Free text is a substring search and can never be an instant, so there the
    // quotes are lexical and come off — without the value changing.
    expect(canonical('"now-24h"')).toBe('now-24h');
  });
});
